### ISSUE TITLE
should KE influence the proposal match code

### ISSUE THREAD

**Issue by cagney** [libreswan/libreswan#1605]
responder has DH1+DH2
initiator sends DH2+DH1;KE=DH2

the proposal code will match DH1 triggering invalid KE when it could match DH2
play with `interop-ikev2-strongswan-14-delete-sa-shared`

**Comment by letoams** [comment]
On Thu, 8 Feb 2024, cagney wrote:

> responder has DH1+DH2
> initiator sends DH2+DH1;KE=DH2
> 
> the proposal code will match DH1 triggering invalid KE when it could match DH2
> play with interop-ikev2-strongswan-14-delete-sa-shared

The proposal list of each peer is proposals it is allowing through local
policy. So to me, it seems the responder should favour the KE used.

But RFC does state that the responder gets to kick, so this is valid if
the responder "strongly favours" another mutually acceptable DH.

**Comment by cagney** [comment]
right.  The RFC suggests the initiator put its prefered proposals firs.  But we do it as the responder.  For instance:

responder has DH1,DH2 -- not +
initiator sends DH2,DH1,KE=DH2

so we clearly prefer DH1 even though initiator put it lower.

Where things get fuzzy is within a proposal.  Our code again has an ordering preference.  It should be weakend a little.

**Comment by cagney** [comment]
An example from `ikev2-invalid-ke-09-preference-ecp`
```
"westnet-eastnet-ipv4-psk-ikev2": IKE SA proposals (connection add):
"westnet-eastnet-ipv4-psk-ikev2":   1:IKE=AES_CBC_256-HMAC_SHA2_256-HMAC_SHA2_256_128-MODP3072
"westnet-eastnet-ipv4-psk-ikev2":   2:IKE=AES_CBC_256-HMAC_SHA2_256-HMAC_SHA2_256_128-ECP_256
...
"westnet-eastnet-ipv4-psk-ikev2" #1: processing IKE_SA_INIT request from 192.1.2.45:UDP/500 containing SA,KE,Ni,N(IKEV2_FRAGMENTATION_SUPPORTED),N(NAT_DETECTION_SOURCE_IP),N(NAT_DETECTION_DESTINATION_IP)
"westnet-eastnet-ipv4-psk-ikev2" #1: proposal 2:IKE=AES_CBC_256-HMAC_SHA2_256-HMAC_SHA2_256_128-MODP3072 chosen from remote proposals 1:IKE:ENCR=AES_CBC_256;PRF=HMAC_SHA2_256;INTEG=HMAC_SHA2_256_128;DH=ECP_256[first-match] 2:IKE:ENCR=AES_CBC_256;PRF=HMAC_SHA2_256;INTEG=HMAC_SHA2_256_128;DH=MODP3072[better-match]
```
The two ends have their proposal order switched, as a result the responder matches ` 2:IKE=AES_CBC_256-HMAC_SHA2_256-HMAC_SHA2_256_128-ECP_256` but then notices it can match its prefered proposal `1:IKE=AES_CBC_256-HMAC_SHA2_256-HMAC_SHA2_256_128-MODP3072`.

**Comment by paulwouters** [comment]
there is different behaviour between:
`ike=aes-sha2;modp3072,aes-sha2;dh19`
and
`ike=aes-sha2;modp3072+dh19`

**Comment by cagney** [comment]
> there is different behaviour between:

Yes, the first contains the two proposals:
```
aes-sha2;modp3072
aes-sha2;dh19
```
and they are in preference order.  The second has only one preference:
```
aes-sha2;modp3072+dh19
```
